### PR TITLE
P2022-448 added authentication and authorization

### PR DESCRIPTION
**Pull request by edych**
- Added new dependencies: Spring Security and Java JWT
- Provided a postman collection 
- Created new Controller for auth related operations
- "/private" can be only accessed with JWT token
- "/public" can be accessed without logging in
- "/register" for register a new Spring User 
- There is also "/login" (default endpoint for Spring Security) 
- After successfully logging in User receives a JWT token 
- Added UsernameTakenException for situations when we are trying to register a User with username that already exists
- Added Authentication and Authorization filters
- Added Security Configuration 
- Added integration Tests for SecuredController
- Adjusted checkstyle

### CODE REVIEW

**Review by Szachista — DISMISSED** [review]

File: src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    @GetMapping("/public")
+    Authentication publicOp() {
+        return SecurityContextHolder.getContext().getAuthentication();
+    }
+
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping("/register")
+    UserDetails register(@RequestParam String username, @RequestParam String password) {
+        UserDetails preparedUser = User.withUsername(username)
```
Comment (patrykzawadzkiintive) [inline src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java:34]: This logic should be moved to new class like UserService.

File: src/main/java/com/intive/patronage22/szczecin/retroboard/exception/UsernameTakenException.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package com.intive.patronage22.szczecin.retroboard.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value= HttpStatus.CONFLICT, reason = "user should not exist")
```
Comment (patrykzawadzkiintive) [inline src/main/java/com/intive/patronage22/szczecin/retroboard/exception/UsernameTakenException.java:6]: Better message: User already exist.

File: postman_collection.json [hunk 3]
```diff
@@ -0,0 +1,105 @@
+{
```
Comment (patrykzawadzkiintive) [inline postman_collection.json:1]: Please move that file to separate direcotry /postman

File: src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        mockMvc
+                .perform(post(url)
+                        .param("username", username)
+                        .param("password", password)
+                        .contentType(MediaType.APPLICATION_FORM_URLENCODED))
+                .andExpect(status().isUnauthorized());
+    }
+
+    @Test
+    void shouldGiveAccessWhenGettingPublicEndpointWithoutAccessToken() throws Exception {
```
Comment (patrykzawadzkiintive) [inline src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java:160]: public_should_return_ok_when_user_annonymous

File: src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        final String url = "/public";
+
+        // then
+        mockMvc.perform(get(url))
+                .andExpect(jsonPath("$.principal").value("anonymousUser"))
+                .andExpect(status().isOk());
+    }
+
+    @Test
+    void shouldNotGiveAccessWhenGettingPrivateEndpointWithoutAccessToken() throws Exception {
```
Comment (patrykzawadzkiintive) [inline src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java:171]: private_should_return_forbidden_when_user_not_logged_in

File: src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    private MockMvc mockMvc;
+
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @MockBean
+    private InMemoryUserDetailsManager inMemoryUserDetailsManager;
+
+    @Test
+    void shouldReturnInJSONsBodyCreatedUserIfItDoesNotExistBefore() throws Exception {
```
Comment (patrykzawadzkiintive) [inline src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java:48]: register_should_return_json_body_with_created_user_data_when_user_not_exist_before

File: src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        verify(inMemoryUserDetailsManager, times(1)).createUser(argumentCaptor.capture());
+        final UserDetails preparedUser = argumentCaptor.getValue();
+
+        assertEquals(username, preparedUser.getUsername());
+        assertNotNull(preparedUser.getPassword());
+        assertNotEquals(password, preparedUser.getPassword());
+    }
+
+    @Test
+    void shouldReturnConflictExceptionIfUserExists() throws Exception {
```
Comment (patrykzawadzkiintive) [inline src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java:84]: register_should_return_confilct_when_user_not_exist

File: src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        mockMvc
+                .perform(post(url)
+                        .param("username", username)
+                        .param("password", password)
+                        .contentType(MediaType.APPLICATION_FORM_URLENCODED))
+                .andExpect(status().isConflict());
+    }
+
+    @Test
+    void shouldReturnAccessTokenWhenLoggingInIfUserExists() throws Exception {
```
Comment (patrykzawadzkiintive) [inline src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java:103]: login_should_return_access_token_when_user_credentials_are_correct

File: src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        String token = map.get("access_token");
+
+        JWTVerifier verifier = JWT.require(Algorithm.HMAC256("secret".getBytes())).build();
+        DecodedJWT decodedJwt = verifier.verify(token);
+
+        assertEquals(username, decodedJwt.getSubject());
+    }
+
+    @Test
+    void shouldReturnUnauthorizedWhenLoggingInIfUserDoesNotExist() throws Exception {
```
Comment (patrykzawadzkiintive) [inline src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java:140]: login_should_return_unauthorized_when_user_not_exist

File: src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+        final String url = "/private";
+
+        // then
+        mockMvc
+                .perform(get(url).contentType(MediaType.APPLICATION_FORM_URLENCODED))
+                .andExpect(status().isForbidden());
+    }
+
+    @Test
+    void shouldGiveAccessWhenGettingPrivateEndpointWithCorrectAccessToken() throws Exception {
```
Comment (patrykzawadzkiintive) [inline src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java:182]: private_should_return_ok_when_user_authenticated

**Review by patrykzawadzkiintive — CHANGES_REQUESTED** [review]

**Comment by patrykzawadzkiintive** [comment]
Try to add final to variables & method parameters wherever you can.

File: src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java (hunk 1 above)
Comment (edych) [reply src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java:34]: Hi! In this case, should I also write unit tests for UserService? Please let me know, thank you :)

File: checkstyle.xml [hunk 11]
```diff
@@ -8,7 +8,7 @@
     <property name="fileExtensions" value="java" />
 
     <module name="LineLength">
-        <property name="max" value="80"/>
+        <property name="max" value="120"/>
```
[the remark this answers is not in the document]
Comment (daniel-kmiecik) [reply checkstyle.xml:11]: I just want to mention that if you are working with a group of people, you should discuss changes that will affect the whole team (before you make them) ;)

File: src/main/java/com/intive/patronage22/szczecin/retroboard/filer/CustomAuthenticationFilter.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+public class CustomAuthenticationFilter extends UsernamePasswordAuthenticationFilter {
+
+    private final AuthenticationManager authenticationManager;
+    private final String jwtSecret;
+
+    @Override
+    public Authentication attemptAuthentication(final HttpServletRequest request, final HttpServletResponse response)
+            throws AuthenticationException {
+
+        final String username = request.getParameter("username");
```
Comment (daniel-kmiecik) [inline src/main/java/com/intive/patronage22/szczecin/retroboard/filer/CustomAuthenticationFilter.java:38]: Both "username" and "password should be passed as an environment variable.

File: src/main/java/com/intive/patronage22/szczecin/retroboard/filer/CustomAuthorizationFilter.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+                    final UsernamePasswordAuthenticationToken authenticationToken =
+                            new UsernamePasswordAuthenticationToken(username, null, authorities);
+
+                    SecurityContextHolder.getContext().setAuthentication(authenticationToken);
+
+                    filterChain.doFilter(request, response);
+                } catch (final Exception exception) {
+                    log.error("Error logging in {}", exception.getMessage());
+
+                    response.setHeader("error", exception.getMessage());
```
Comment (daniel-kmiecik) [inline src/main/java/com/intive/patronage22/szczecin/retroboard/filer/CustomAuthorizationFilter.java:66]: Why do you put this in header?
It's unneccessary when you are passing an error message into response object

File: src/main/java/com/intive/patronage22/szczecin/retroboard/filer/CustomAuthorizationFilter.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+                    response.setStatus(FORBIDDEN.value());
+                    response.setContentType(APPLICATION_JSON_VALUE);
+
+                    final Map<String, String> error = new HashMap<>();
+                    error.put("error_message", exception.getMessage());
+                    new ObjectMapper().writeValue(response.getOutputStream(), error);
```
Comment (daniel-kmiecik) [inline src/main/java/com/intive/patronage22/szczecin/retroboard/filer/CustomAuthorizationFilter.java:72]: Spring creates default ObjectMapper, so there's no need to create a new instance. Just try to autowire objectmapper in this class. Should work until you need to customize it.

File: checkstyle.xml (hunk 11 above)
[the remark this answers is not in the document]
Comment (edych) [reply checkstyle.xml:11]: Of course, you are absolutely right. Then maybe let's talk about it all together at our call today? :)

**Review by patrykzawadzkiintive — DISMISSED** [review]

**Review by golldfish — DISMISSED** [review]

File: postman/postman_collection.json [hunk 15]
```diff
@@ -0,0 +1,105 @@
+{
```
Comment (daniel-kmiecik) [inline postman/postman_collection.json:1]: Why do you want to commit this file to the repo? It's something like additional tool, it's not a part of our service.

File: src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.HttpStatus;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequiredArgsConstructor
+public class SecuredController {
```
Comment (daniel-kmiecik) [inline src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java:13]: The name of the class should refer to its business logic.

File: src/main/java/com/intive/patronage22/szczecin/retroboard/exception/UsernameTakenException.java [hunk 17]
```diff
@@ -0,0 +1,12 @@
+package com.intive.patronage22.szczecin.retroboard.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value= HttpStatus.CONFLICT, reason = "User already exist")
+public class UsernameTakenException extends RuntimeException {
+
+    public UsernameTakenException() {
```
Comment (daniel-kmiecik) [inline src/main/java/com/intive/patronage22/szczecin/retroboard/exception/UsernameTakenException.java:9]: It should take a message parameter or a static message variable to be passed to the super() constructor.
Now we don't know what is the cause of the exception.

File: src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+
+        // when
+        when(userService.register(username, password)).thenThrow(new UsernameTakenException());
+
+        // then
+        mockMvc
+                .perform(post(url)
+                        .param("username", username)
+                        .param("password", password)
+                        .contentType(MediaType.APPLICATION_FORM_URLENCODED))
```
Comment (daniel-kmiecik) [inline src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java:100]: Our controllers should produces JSON format.

File: src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+    private final UserService userService;
+
+    @GetMapping("/private")
+    Authentication privateOp() {
+        return SecurityContextHolder.getContext().getAuthentication();
+    }
+
+    @GetMapping("/public")
```
Comment (daniel-kmiecik) [inline src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java:22]: Currently, the public endpoint is /boards, so we do not need another one only for tests. Correct me if I am wrong.

File: src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+                .andExpect(status().isConflict());
```
Comment (daniel-kmiecik) [inline src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java:101]: You also should verify the message of the error.

File: src/main/java/com/intive/patronage22/szczecin/retroboard/filer/CustomAuthorizationFilter.java (hunk 14 above)
Comment (edych) [reply src/main/java/com/intive/patronage22/szczecin/retroboard/filer/CustomAuthorizationFilter.java]: done

File: postman/postman_collection.json (hunk 15 above)
Comment (edych) [reply postman/postman_collection.json]: deleted

File: src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java (hunk 16 above)
Comment (edych) [reply src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java]: renamed

File: postman/postman_collection.json (hunk 15 above)
Comment (edych) [reply postman/postman_collection.json]: done

File: src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java (hunk 16 above)
Comment (edych) [reply src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java]: done

File: src/main/java/com/intive/patronage22/szczecin/retroboard/exception/UsernameTakenException.java [hunk 21]
```diff
@@ -0,0 +1,12 @@
+package com.intive.patronage22.szczecin.retroboard.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value= HttpStatus.CONFLICT, reason = "User already exist")
+public class UsernameTakenException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (edych) [reply src/main/java/com/intive/patronage22/szczecin/retroboard/exception/UsernameTakenException.java]: yes :) done

File: src/main/java/com/intive/patronage22/szczecin/retroboard/service/UserService.java [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+    private final PasswordEncoder passwordEncoder;
+
+    public UserDetails register(final String username, final String password) {
+        final UserDetails preparedUser = User.withUsername(username)
+                .password(password)
+                .roles("USER")
+                .passwordEncoder(passwordEncoder::encode)
+                .build();
+
+        if (userExists(username)) {
```
[the remark this answers is not in the document]
Comment (edych) [reply src/main/java/com/intive/patronage22/szczecin/retroboard/service/UserService.java]: done

File: src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java (hunk 19 above)
Comment (edych) [reply src/main/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredController.java]: correct. I have deleted this enpoint

File: src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java (hunk 20 above)
Comment (edych) [reply src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java]: done

File: src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java (hunk 18 above)
Comment (edych) [reply src/test/java/com/intive/patronage22/szczecin/retroboard/controller/SecuredControllerTest.java]: done

File: src/main/java/com/intive/patronage22/szczecin/retroboard/security/WebSecurityConfig.java [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    protected void configure(final AuthenticationManagerBuilder auth) throws Exception {
+        auth.userDetailsService(userDetailsService).passwordEncoder(passwordEncoder);
+    }
+
+    @Override
+    protected void configure(final HttpSecurity http) throws Exception {
+        http.csrf().disable();
+        http.sessionManagement().sessionCreationPolicy(STATELESS);
+        http.authorizeRequests().antMatchers("/register", "/boards", "/public").permitAll();
```
Comment (daniel-kmiecik) [inline src/main/java/com/intive/patronage22/szczecin/retroboard/security/WebSecurityConfig.java:46]: There's no more "/public" endpoint.

File: src/main/java/com/intive/patronage22/szczecin/retroboard/service/UserService.java [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+
+        inMemoryUserDetailsManager.createUser(preparedUser);
+
+        final User createdUser = (User) inMemoryUserDetailsManager.loadUserByUsername(preparedUser.getUsername());
+        createdUser.eraseCredentials();
+
+        return createdUser;
+    }
+
+    private boolean userExists(final String username) {
```
Comment (daniel-kmiecik) [inline src/main/java/com/intive/patronage22/szczecin/retroboard/service/UserService.java:41]: Methods which return boolean should start with "is". In this case should be "isUserExist".

Comment (Szachista) [reply src/main/java/com/intive/patronage22/szczecin/retroboard/service/UserService.java:41]: isUserExisting (?)

File: src/main/java/com/intive/patronage22/szczecin/retroboard/security/WebSecurityConfig.java (hunk 23 above)
Comment (edych) [reply src/main/java/com/intive/patronage22/szczecin/retroboard/security/WebSecurityConfig.java:46]: Sorry, I've missed that. Done

File: src/main/java/com/intive/patronage22/szczecin/retroboard/service/UserService.java (hunk 24 above)
Comment (edych) [reply src/main/java/com/intive/patronage22/szczecin/retroboard/service/UserService.java:41]: done

**Review by golldfish — APPROVED** [review]

**Review by daniel-kmiecik — APPROVED** [review]

**Pull request merged**